### PR TITLE
Clean up and refactor array element access related code

### PR DESCRIPTION
**Pull request by sifmelcara**
### Description

This is factored out from #6881 , as #6881 probably become too big to be efficiently reviewed.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

**Comment by sifmelcara** [comment]
Motivation of moving those code to separate function: https://github.com/ethereum/solidity/pull/6881#discussion_r289810690

File: libsolidity/codegen/ArrayUtils.cpp [hunk 1]
```diff
@@ -1055,28 +1055,27 @@ void ArrayUtils::accessIndex(ArrayType const& _arrayType, bool _doBoundsCheck, b
 	switch (location)
 	{
 	case DataLocation::Memory:
+		// stack: <base_ref> <index>
+		if (!_arrayType.isByteArray())
+			m_context << u256(_arrayType.memoryHeadSize()) << Instruction::MUL;
```
Comment (sifmelcara) [inline libsolidity/codegen/ArrayUtils.cpp:1060]: (We may want to change this to `_arrayType.baseType()->memoryHeadSize()`, as discussed at https://github.com/ethereum/solidity/pull/6881#discussion_r289808363)

Comment (chriseth) [reply libsolidity/codegen/ArrayUtils.cpp:1060]: Yes, it should be ` _arrayType.baseType()->memoryHeadSize()`

File: libsolidity/codegen/ContractCompiler.cpp [hunk 2]
```diff
@@ -893,9 +899,9 @@ bool ContractCompiler::visit(VariableDeclarationStatement const& _variableDeclar
 
 	// Local variable slots are reserved when their declaration is visited,
 	// and freed in the end of their scope.
-	for (auto _decl: _variableDeclarationStatement.declarations())
-		if (_decl)
-			appendStackVariableInitialisation(*_decl);
+	for (auto decl: _variableDeclarationStatement.declarations())
```
Comment (chriseth) [inline libsolidity/codegen/ContractCompiler.cpp:902]: I think it was `_decl` because `decl` might be a reserved word on some systems.

**Review by chriseth — APPROVED** [review]
As a refactor, this is fine to merge. We maybe should use a new PR for the fix.

**Pull request merged**